### ISSUE TITLE
Rendered content of `<For/>` stops updating when underlying signal changes

### ISSUE THREAD

**Issue by LoganWalls** [leptos-rs/leptos#1289]
**Describe the bug**
I am trying to render a dynamic list of elements using `<For/>`. In this example, I have a search box that searches over the names of countries, and a `<For/>` that renders each matched country as a button:

```rust
let all_items = create_resource(cx, || (), fetch_items);
let visible_items = ||{
  // ...
  filtered_copy_of_all_items
}

// ...
view!{cx, <For
    each=visible_items
    key=|item| item.id
    view=move |cx, item| {
      view! {
        cx,
        <button>{item.data.key}</button>
      }
    }
  />
}
```

The items in `all_items` are fixed (it's a resource that is fetched only once), and each `Item` has an id field that is a unique `usize` which never changes.

An example that I can easily reproduce is searching for "Canada". The contents of the `<For/>` update correctly for at least two updates. But at some point the `<For/>` stops updating. As you can see in this gif, even though I change the query to "United", the results do not change at all. However, if I make a side-effect that console logs `visible_items`, I can see that the signal itself is updating correctly. For example, in the gif you can see that after I change the query to "United", `visible_items` contains the UK and US as one would expect.

**Leptos Dependencies**
```toml
leptos = { version = "0.4", features = ["nightly", "csr"] }
```

**To Reproduce**
1. Clone the example repository [here](https://github.com/LoganWalls/leptos-mwe)
2. Change directories into the cloned repository
3. Run `trunk serve`, and open the url it prints in your browser
4. Type "Canada" into the search box
5. Delete "Canada" and try typing another query like "United"

**Expected behavior**
I would expect the items rendered in the `<For each=visible_items />` tag to match the items in the `visible_items` signal.

**Screenshots**
![leptos](https://github.com/leptos-rs/leptos/assets/2934282/9d25200f-f259-4488-9326-9354ee7fe0c3)

Thank you in advance for your help!

**Comment by gbj** [comment]
Thanks! This definitely appears to be path-dependent — i.e., if I paste in "Canada" and then start deleting letters, it works, but if I type "Canada", and then start deleting, it sticks. It is likely related to the issue with moving multiple items in a single change in #1274.

For your particular use case, by the way, where it's only 10 rows and often completely removing all the rows — I almost wonder if a plain iterator might work better anyway. 
```rust
move || {
    visible_items()
        .into_iter()
        .map(|item| {
            view! { cx, <button>{item.data.key}</button> }
        })
        .collect_view(cx)
```
The reasoning here would be that keyed lists are mostly useful when you're only adding/subtracting one item at a time, in a large list, or when you're moving items around and don't want to rerender. When you're only showing 10 items and rendering just a button per row, with many of the items swapping out on every change anyway, it may really not make a difference!

Clearly this is still a bug we'll need to figure out, of course.

**Comment by LoganWalls** [comment]
Great, thank you so much! That works well as a work-around for now. 

Some of my actual use cases may be ultimately better suited for the `<For/>` version, so I'll definitely keep an eye on this issue. But this works fine for now and I'm not noticing any performance issues, so I don't mind waiting (and I'm sure you have a lot on your plate).

Thanks again for all of your hard work on this project!

**Comment by gbj** [comment]
Looks like we'll be fixing this sooner rather than later. The linked PR #1296 should make this behave exactly correctly. We're just adding one more small optimization and it should be good to go.

**Comment by LoganWalls** [comment]
Awesome, thank you so much!

**Comment by gbj** [comment]
This is fixed by #1296.
